src/installer/corehost/cli/apphost/static/CMakeLists.txt

```diff
@@ -214,7 +244,25 @@ if(CLR_CMAKE_USE_SYSTEM_LIBUNWIND)
     )
 endif()
 
+if(CLR_CMAKE_TARGET_LINUX OR CLR_CMAKE_TARGET_FREEBSD OR CLR_CMAKE_TARGET_NETBSD OR CLR_CMAKE_TARGET_SUNOS)
+    # These options are used to force every object to be included even if it's unused.
+    set(START_WHOLE_ARCHIVE -Wl,--whole-archive)
+    set(END_WHOLE_ARCHIVE -Wl,--no-whole-archive)
+endif(CLR_CMAKE_TARGET_LINUX OR CLR_CMAKE_TARGET_FREEBSD OR CLR_CMAKE_TARGET_NETBSD OR CLR_CMAKE_TARGET_SUNOS)
+
+if(CLR_CMAKE_TARGET_OSX)
+    # These options are used to force every object to be included even if it's unused.
+    set(START_WHOLE_ARCHIVE -force_load)
+    set(END_WHOLE_ARCHIVE )
+endif(CLR_CMAKE_TARGET_OSX)
+
+set_property(TARGET singlefilehost PROPERTY ENABLE_EXPORTS 1)
+
 target_link_libraries(
     singlefilehost
     ${NATIVE_LIBS}
+
+    ${START_WHOLE_ARCHIVE}
+    ${RUNTIMEINFO_LIB}
```

Comment:
From the latest installer build, it seems like we have dynamic linkage on linux in statically linked sinlgefilehost (although overrider counterparts are present and should be used instead?):

```sh
$ ldd -r ~/.dotnet6/packs/Microsoft.NETCore.App.Host.linux-x64/6.0.0-preview.2.21105.12/runtimes/linux-x64/native/singlefilehost 
	linux-vdso.so.1 (0x00007fffd091a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fa96ea21000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fa96e81d000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fa96e600000)
	libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007fa96e3b5000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fa96e1ad000)
	libanl.so.1 => /lib/x86_64-linux-gnu/libanl.so.1 (0x00007fa96dfa9000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fa96dc20000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fa96d882000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fa96d66a000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa96d279000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa96f66f000)
	libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007fa96cfa3000)
	libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007fa96cd71000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007fa96cb6d000)
	libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007fa96c962000)
	libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007fa96c75e000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007fa96c544000)
```
is this correct that dynamic linkages are not stripped?

Reply:
@am11 do you mean all the libraries listed above? If it is the case, we don't want to statically link 3rd party libraries. The reason is that this way, we get potential fixes to those libraries through the repo and we don't need to release updated host every time one of those get some important fix. In some of them, I think it would be even a licensing problem.

Reply:
@janvorli, yup, it looks like libunwind and libsssl are statically linked in the binary. When I read about "super host" in the design doc: https://github.com/dotnet/designs/blob/5a6bf8fef1b072a7ed578a23dbd3d2ee7dd89d11/accepted/2020/single-file/design.md#host-builds, I was under the impression that:

```sh
$ file singlefilehost | fold -72
singlefilehost: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Lin
ux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildI
D[sha1]=27997343859af70e0bfc5e50a4d1059b3513a1c4, for GNU/Linux 2.6.32, 
stripped, too many notes (256)
```
will just become
```sh
$ file singlefilehost | fold -72
	not a dynamic executable
```
Maybe objects file could be provided so folks can statically link their apps?

Reply:
libssl is not statically linked, but we load it dynamically, so it is not on the direct dependencies list. As for providing object files, that sounds like an interesting idea. However, linking e.g. standard c library statically is almost never a good idea. The problem is that a 3d party shared library used in some user code DllImport would most likely be dynamically linked to libc and mixing two libcs in one process is a recipe for problems.

Reply:
For reference, in node.js this is possible:

```sh
$ curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.37.2/install.sh | bash
$ nvm install 12.9.1
$ npm install nexe
$ echo 'console.log("Hello World!");' > hi.js

# static
$ nexe -t static hi.js
$ file hi | fold -72
hi: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically link
ed, with debug_info, not stripped

# dynamic
$ nexe hi.js
$ file hi | fold -72
hi: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), dynamicall
y linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32,
 BuildID[sha1]=3cb1a5d531f2b89baef69d5bbb4f0688959a572e, with debug_info
, not stripped
```

Reply:
I know, I have noticed this a long time ago when I was enabling portable runtime builds. Their repo contains copies of source all of their dependencies except of standard c/c++ libraries (about 16 libraries). I guess that they took the risk of possibility of having multiple standard C libraries in the same process.
This is an interesting write-up I've used as a source of details on creating portable binaries for Linux at that time:
https://insanecoding.blogspot.com/2012/07/creating-portable-linux-binaries.html

Reply:
Thank you for sharing the link! :)

Ultimately I am trying to build a bootable image out of singlefilehost linked to a new cloud-oriented unikernel nanos (utsname->sysname: 'pugnix'): https://github.com/nanovms/nanos/blob/9b43cfcb6dedad4c896c4efc2173336c4f83f71a/src/unix/syscall.c#L1527-L1529. It is basically a light-weight implementation of Linux kernel interface, with stubbed out syscalls and libc ([klib](https://attractivechaos.github.io/klib)). The produced .img file out of user app has a slim bootloader, that can be burned to a bootable device as well as deployed to cloud: HyperV in Azure, run with qemu etc. Basically the concept is: "your one (and only one) app _is_ the OS" (almost) so there is very limited surface area for vulnerabilities (and other positive/negative virtues of a general-purpose OS kernel). There was am attempt for .NET5 (and also .NET6) singlefile app: https://github.com/nanovms/ops-examples/issues/47 but scenario is not fully functional yet (that bootable myapp1.img is 109M; OS+silefileapp+deps).

The ideal situation for app for unikernel environment is: it is statically linked and has embedded (native) debug symbols (`.debug_info` and other `.debug_*` ELF sections instead of `.gnu_debuglink` pointing to `.dbg` file, which we currently have). I think that can be achieved if a separate runtime pack provides the object files and a recipie to embed debug symbols (for app built with Debug configuration). However, dynamic binaries work but with `--debug` switch in their tooling, fails [these two](https://github.com/nanovms/ops/blob/bf36f3c1a4719ec444daf6a8c38b7fb49497139f/cmd/flags_start_image.go#L54-L60) validations. Therefore, I opted for `--trace` option to track down coreclr_initialize failures.